### ISSUE TITLE
nested modules implicit aliasing

### ISSUE THREAD

**Issue by tim2CF** [elixir-lang/elixir#9191]
Hello!
I found some differences between Elixir `1.8.2` and `1.9.0` compilers. Looks like in new version local references to nested modules are not working anymore:

```elixir
defmodule Foo do
  defmodule Elixir.Foo.Bar do
    def hello, do: "hello"
  end

  @world "#{Bar.hello()}, world!"
  def world, do: @world
end
```

This is minimal example which compiles in `1.8.2` but not compiles in `1.9.0`.
This particular one is not useful and looks weird, but I used similar functionality a lot in metaprogramming.
For example how looks transition from `1.8.2` to `1.9.0` for one of my libraries:
https://github.com/coingaming/gen_enum/commit/3a1ad5cf94a30790a360147a6980574046147933

I’m not sure is this bug or feature, but this is significant difference in behaviour. In some cases this implicit local module aliasing was very useful. Anyway, if this behaviour is expected - this is breaking change and should be described in release notes more detailed.

**Comment by josevalim** [comment]
Hi @tim2CF,

This was considered a bug fix because `Elixir.Foo.Bar` was always meant to opt out of the aliasing mechanism. You can keep the alias though if you define the module name as `defmodule Bar do`. Is there a reason why you don't do so?

In any case, you are absolutely correct, we were not expecting the bug fix to affect anyone, so we will be clearer about the potential downsides of it in the CHANGELOG.

**Comment by josevalim** [comment]
And here is the commit with the bug fix and rationale: a6e9de56dbd5062e253488df6a0eb3b033d19dcd

**Comment by tim2CF** [comment]
The context is generic macro for enums which work consistently in different use cases:

- outside of the module
- inside the module without namespace nesting
- inside the module with namespace nesting

(links are showing different definitions but same behaviour)
https://github.com/coingaming/gen_enum/blob/06af64b1d892c222a23155829360c53ff861188d/test/gen_enum_test.exs#L9-L32
https://github.com/coingaming/gen_enum/blob/06af64b1d892c222a23155829360c53ff861188d/test/gen_enum_test.exs#L38-L43

This macro is a bit complex and generates 3-4 nested modules with other macros under given namespace, and it is much easier to do code generation under global namespace which is propagated to macro's private functions just as single argument: 

https://github.com/coingaming/gen_enum/blob/06af64b1d892c222a23155829360c53ff861188d/lib/gen_enum.ex#L104

I already fixed this with additional explicit alias generator which "emulates" `1.8.2` compiler feature

https://github.com/coingaming/gen_enum/blob/06af64b1d892c222a23155829360c53ff861188d/lib/gen_enum.ex#L106-L116

Just wanted to confirm that this was not a bug
